cpu/atmega2560/periph/spi.c

```diff
@@ -0,0 +1,235 @@
+#endif
+#if SPI_1_EN
+    [SPI_1] = MUTEX_INIT,
+#endif
+};
+
+int spi_init_master(spi_t dev, spi_conf_t conf, spi_speed_t speed)
+{
+    uint8_t _spcr = (1<<SPI_0_MSTR)|(1<<SPI_0_SPE);
+
+    // Assume F_CPU = 16.000.000 Hz
+    switch (speed) {
+        case SPI_SPEED_100KHZ:
+            // Actual rate = 125.000 Hz
+            _spcr |= (1<<SPI_0_SPR1)|(1<<SPI_0_SPR0);
+            break;
+        case SPI_SPEED_400KHZ:
+            // Actual rate = 250.000 Hz
+            _spcr |= (1<<SPI_0_SPR1);
+            break;
+        case SPI_SPEED_1MHZ:
+            // Actual rate = 1.000.000 Hz
+            _spcr |= (1<<SPI_0_SPR0);
+            break;
+        case SPI_SPEED_5MHZ:
+            // Actual rate = 4.000.000 Hz
+            _spcr |= 0;
+            break;
+        case SPI_SPEED_10MHZ:
+            // Actual rate = 8.000.000 Hz
+            _spcr |= (1<<SPI_0_SPI2X);
+            break;
+    }
+
+    switch (conf) {
+        case SPI_CONF_FIRST_RISING:
+            // Do nothing for CPOL=0, CPHA=0
+            break;
+        case SPI_CONF_SECOND_RISING:
+            _spcr |= (1<<SPI_0_CPHA);
+            break;
+        case SPI_CONF_FIRST_FALLING:
+            _spcr |= (1<<SPI_0_CPOL);
+            break;
+        case SPI_CONF_SECOND_FALLING:
+            _spcr |= (1<<SPI_0_CPOL)|(1<<SPI_0_CPHA);
+            break;
+    }
+
+    switch (dev) {
+#if SPI_0_EN
+        case SPI_0:
+            SPI_0_DDR |= (1<<SPI_0_MOSI)|(1<<SPI_0_SCK);
+            SPI_0_DDR &= ~((1<<SPI_0_MISO));
+            SPI_0_CR = _spcr;
+            return 0;
+            break;
+#endif
+    }
+    return -2;
+
+}
+
+int spi_init_slave(spi_t dev, spi_conf_t conf, char (*cb)(char data)) {
+    uint8_t _spcr = (1<<SPI_0_SPE);
+
+    switch(conf) {
+        case SPI_CONF_FIRST_RISING:
+            // Do nothing for CPOL=0, CPHA=0
+            break;
+        case SPI_CONF_SECOND_RISING:
+            _spcr |= (1<<SPI_0_CPHA);
+            break;
+        case SPI_CONF_FIRST_FALLING:
+            _spcr |= (1<<SPI_0_CPOL);
+            break;
+        case SPI_CONF_SECOND_FALLING:
+            _spcr |= (1<<SPI_0_CPOL)|(1<<SPI_0_CPHA);
+            break;
+    }
+
+    switch(dev) {
+#if SPI_0_EN
+        case SPI_0:
+            SPI_0_DDR |= (1<<SPI_0_MISO);
+            SPI_0_DDR &= ~((1<<SPI_0_MOSI)|(SPI_0_SCK));
+            SPI_0_CR = _spcr;
+            config[dev].rx_cb = cb;
+            return 0;
+            break;
+#endif
+    }
+
+    return -1;
+}
+
+int spi_conf_pins(spi_t dev) {
+    // Done by init functions
+    return 0;
+}
+
+int spi_acquire(spi_t dev) {
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    mutex_lock(&locks[dev]);
+    return 0;
+}
+int spi_release(spi_t dev) {
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int spi_transfer_byte(spi_t dev, char out, char *in) {
+    switch (dev) {
+#if SPI_0_EN
+    case SPI_0:
+        SPI_0_DR = out;
```

Comment:
I'm not a ATmega expert, but wouldn't it be better to check if a previously set data register was already written out?